src/System.Buffers/src/System/Buffers/ByteBufferPool.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    public sealed class ByteBufferPool : IBufferPool<byte[]>
+    {
+        private const int NUMBER_OF_BUFFERS_IN_BUCKET = 50;
+        private int _maxBufferSize;
+        private BufferBucket[] _buckets;
+
+        // 2MB taken as the default since the average HTTP page is 1.9MB
+        // per http://httparchive.org/interesting.php, as of October 2015
+        public ByteBufferPool(int maxBufferSizeInBytes = 2048000)
+        {
+            int maxBuckets = SelectBucketIndex(maxBufferSizeInBytes);
+
+            _maxBufferSize = maxBufferSizeInBytes;
+            _buckets = new BufferBucket[maxBuckets + 1];
+
+            for (int i = 0; i < _buckets.Length; i++)
+            {
+                _buckets[i] = new BufferBucket(GetMaxSizeForBucket(i), NUMBER_OF_BUFFERS_IN_BUCKET);
+            }
+        }
+
+        public byte[] RentBuffer(int size, bool clearBuffer = false)
+        {
+            byte[] buffer;
+            if (size > _maxBufferSize)
+            {
+#if DEBUG
+                System.Diagnostics.Debugger.Break();
+#endif
+                buffer = new byte[size];
+            }
+            else
+            {
+                buffer = _buckets[SelectBucketIndex(size)].Rent();
+                if (clearBuffer) Array.Clear(buffer, 0, buffer.Length);
```

Comment:
I wonder if Return should not have clear either or only. I think clear on return is more useful for code that does not want to leak secrets.